Hover tooltips now fetch full docs with racer fallback

aloucks opened the pull request:
Nearly all code sections now load tooltips. Traits, enums, structs, and unions are extracted from source in order to capture additional type information (e.g. generics and trait bounds). Function signatures and the aforementioned items above are also formatted with `rustfmt`.

Documentation is extracted from the source and post-processed so that code block markdown is formatted correctly by vscode. For example, annotated code blocks with `no_run` or `ignore` qualifiers would previously be formatted as plaintext. 

The context of local variable declarations is now also included in the tooltip, in addition to the type information. 

In the event of source extraction failure or racer errors, the type information and docs provided by the save analysis are used. 

Documentation URLs and each section of the hover tooltip now have at least one line between them and no longer run together. 

![hover](https://user-images.githubusercontent.com/221559/42406128-ba4f4ec2-816f-11e8-92c7-c671cb5ec1a3.gif)

alexheretic commented on src/actions/hover.rs:
```diff
@@ -0,0 +1,480 @@
+                }
+                lines.lines().filter(|line| line.trim() != "").map(|line| {
+                    format!("{}\n", &line[config.tab_spaces()..])
+                }).collect()
```
Is the trailing '\n' wanted for vscode here, it causes unsightliness in atom

![before](https://user-images.githubusercontent.com/2331607/42566639-16faba7a-84ff-11e8-8b31-f739f039e649.png)

Removing the trailing newline fixes this
```rust
...
let mut s: String = lines
    .lines()
    .filter(|line| !line.trim().is_empty())
    .map(|line| format!("{}\n", &line[config.tab_spaces()..]))
    .collect();
s.pop();
s
```
![after](https://user-images.githubusercontent.com/2331607/42566653-1f815dd4-84ff-11e8-831f-e15bd8e8e526.png)

alexheretic commented:
Thanks this will be a great addition.

### Newline spacing
I'm not sure about the `\n\n` bits, are these just for vscode? Could this be fixed _in_ vscode? This could cause other clients issue. On the other hand, whilst atom does not benefit from the spacing it does neem to actually ignore them mostly so I guess I don't care too much. Maybe other client maintainers would like to chime in?

An exception I noticed is function signature trailing `\n` did cause issue in atom, _comments inline_.

In general _if_ this is an issue we could always make the spacing hacks configurable.

### Config
I had to dig a bit to figure out what the `full_docs` config actually did. But I get it now from the `rls-data` docs

> Include all documentation for items. (If false, only includes the summary (first paragraph) for each item).

Can you add documentation for this config to the readme?

I assume this config is still relevant to some clients? In atom I'll just be setting this to `true` as atom has scrolling boxes for this stuff.

### Implementation
I'm a bit worried the `hover.rs` code could be fragile. Could the manual string checking and replacing have issues down the line? Perhaps some unit testing would garner more confidence & make it easier to maintain.

You should run rustfmt on the code.

alexheretic commented on src/actions/hover.rs:
```diff
@@ -0,0 +1,480 @@
+            let ty = racer_ty.unwrap_or(hover_span_ty);
+            let ty = ty.trim();
+            if !ty.is_empty() {
+                contents.push(MarkedString::from_language_code("rust".into(), format!("{}", ty)));
```
Minor, but the `format!("{}", foo_str)`  usages could just be `foo_str.into()` / `to_owned()`

aloucks replied:
I think at one point there was newline padding in here as well. I agree with your above comments that adding extra newlines isn't so wonderful. It is definitely to fix issues with the way it's rendered in vscode. I almost left a comment in the source. 

I have been trying to figure out why rls-vscode does not include line separators (between each element of the `HoverContents::Array(Vec<MarkedString>)` data. I've looked at the hover providers for both the Go extension and the other [rust extension](https://github.com/editor-rs/vscode-rust/blob/master/src/components/completion/completion_manager.ts#L278) and I haven't been able to figure out why rls-vscode is behaving differently. Both of those extensions have explicit hoverProviders in typescript, but as far as I can tell the logic is the same. 

Note the thin black lines between the defintion, doc url, and the docs (the screen shot is using the old [rust extension](https://github.com/editor-rs/vscode-rust/blob/master/src/components/completion/completion_manager.ts#L278) ). For some reason those lines are missing with the [rls-vscode](https://github.com/rust-lang-nursery/rls-vscode) extension. 

![image](https://user-images.githubusercontent.com/221559/42572590-c185df46-84e7-11e8-8cb5-18604aa47cc1.png)

I've been trying to figure this out for a few days. Any help would be greatly appreciated!

Thanks for the review. I'll take a look at the remaining items later this evening.

alexheretic replied:
Well vscode isn't my area & atom handles this ok, _ie has the line separators_.
![box-hover](https://user-images.githubusercontent.com/2331607/42573068-006b4658-8513-11e8-9ae1-2494dc5eb1ba.png)

I'd simply raise it as a vscode issue & be fairly satisfied that rls is doing the right thing, but I appreciate you want to have it looking right there off the bat.

alexheretic commented on src/actions/hover.rs:
```diff
@@ -0,0 +1,480 @@
+                    lines = lines[0..back_pos].into();
+                }
+                lines.lines().filter(|line| line.trim() != "").map(|line| {
+                    format!("{}\n", &line[config.tab_spaces()..])
```
Perhaps we should explicitly check that `&line[config.tab_spaces()..]` will work in the case rustfmt acts strangely on the dummy input.

aloucks commented:
In regards to `full_docs` -- this was my original intended change (#927). Although, that option alone doesn't quite cut it. I added some inline comments in the source as to why it's not very useful right now. The std library doesn't have it enabled when it's compiled and it seemed like even when I turned it on for my project and dependencies, it didn't always work (although that could have been something flaky). 

Given that the source extraction overrides any docs that come out of the save analysis right now, do you want me to still add notes to the README?

I cleaned up the code around parsing the rustfmt output. It should be robust now. I'm still trying to figure out what's going on in the rls-vscode extension that's causing the line separators to disappear. I'm going to continue investigating this for a while longer before I commit to a decision with the newlines.

aloucks commented:
The line separator issue is fixed by upgrading vscode-languageclient.

https://github.com/rust-lang-nursery/rls-vscode/pull/367

aloucks commented:
@alexheretic I think this is ready now.

All of the extra newlines for vscode formatting have been removed. The `full_docs` option now has some information in the `README` and everything has been refactored and cleaned up.

All of the source extraction and parsing now have unit tests. There's also integration tests that record the tooltip output into `target/tests/hover/save_data`. The output is then compared against `test_data/hover/save_data`. https://github.com/aloucks/rls/tree/hover/test_data/hover

alexheretic commented on src/actions/hover.rs:
```diff
@@ -0,0 +1,1691 @@
+            }
+        },
+        Err(e) => {
+            error!("format_method: error: {:?}", e);
```
The error could tell us more when the `impl Dummy` code injection fails, ie
```rust
error!(
    "format_method: {:?} {:?}",
    format!("impl Dummy {{ {} {{ unimplmented!() }} }}", the_type),
    e,
);
```

_Should deduplicate the format! logic though, to an inner function perhaps._

Ie an error I encountered it's clearer from the dummy-code what's going wrong
`ERROR 2018-07-18T10:36:49Z: rls::actions::hover: format_method: "impl Dummy { fn matrix_something(\n_a: [[f32; 4] { unimplmented!() } }" (ParseError, Summary { has_operational_errors: false, has_parsing_errors: true, has_formatting_errors: false, has_macro_format_failure: false, has_check_errors: false, has_diff: false, timer: Initialized(Instant { tv_sec: 6194, tv_nsec: 566976823 }) })`

alexheretic commented:
Nice work!

Here is another test case I found, the one-per-line formatted 2d arrays are causing mischief.
```rust
/// Do nothing with matices.
fn matrix_something(
    _a_matrix: [[f32; 4]; 4],
    _b_matrix: [[f32; 4]; 4],
    _c_matrix: [[f32; 4]; 4],
    _d_matrix: [[f32; 4]; 4],
) {}
```

I added a comment on improving the `format_method` error messages, as I wouldn't be surprised if more cases turn up.

Some more suggestions
* Run rustfmt on hover.rs / any other new code
* Run clippy there are lots of minor issues that may be of interest.

aloucks commented:
@alexheretic  `format_object` and `format_method` now log the input string on error. The issue was actually with the extraction of the method declaration, not the formatting. I've fixed this and added add tests case for both. 

All clippy issues in new code are addressed and corrected. I ran `cargo fmt` and it re-wrote the entire project which I didn't think would be appreciated in a code review :) I backed that out and just ran rustfmt on `hover.rs`.

alexheretic commented:
Good stuff! Lgtm. @nrc @Xanewok what do you guys think, merge?

aloucks commented:
There are multiple benefits here:

1. The tooltips now have full documentation.
1. Locals, statics, and consts shows additional context around how they were declared.
1. The generic type information is now presented in structs/traits, and function signatures. 
1. Function signatures and structs/traits are now formatted with `rustfmt` and look much nicer when they're long.
1. Many items that had no tooltip at all before, now have tooltips because of racer. For example `std::thread::spawn` now shows the module docs when you hover over `thread` and hovering over `Vec` in `Vec::new()` shows the docs for `Vec`.

Reading the docs from a single file on demand doesn't have any noticeable performance difference. It feels just as fast as it did before. Reading a small block of contiguous data from a single file is not slow. The `Vfs` is also caching it as well. Racer is quite fast too, which is surprising. 

The fallback on racer right now isn't really for docs, per say. It's for when there's no `Def` available (and thus not enough information to extract the docs or function/struct definitions).

>  one idea is to move the docs out of the save-analysis files and into their own files which we load asynchronously.

I was thinking about this too and I think it's a solid idea. I started looking into how the compiler extracts save analysis. The challenge is that even with the docs external, we're still not getting tooltips on many items. 

I think the PR as it exists is a good stepping stone and improvement. The dependency on racer is rather contained. As the save analysis improves (e.g. a `Def` becomes available in more contexts), it will simply fall back on racer for fewer items.

alexheretic commented:
Regarding racer fallback we should probably disable it when config `racer_completion` is disabled, and updated he config docs too. That config is for projects where racer enters an unending loop, or otherwise performs very badly. So I imagine hover racer queries would similarly suffer there.

alexheretic commented:
There is a bit of duplicated functionality with _textDocument/definition_ now, though the formatted one in hover can actually be more useful sometimes. 

In atom if you ctrl-hover you get to see the definition along with the type and docs. Which makes sense as ctrl-click would take you there. This has some overlap with the new functionality.

## Function example
#### Hover (before)
![hover-before](https://user-images.githubusercontent.com/2331607/42935700-36c00204-8b42-11e8-9bd1-23be59846b9b.png)
#### Ctrl-hover (before)
![ctrl-hover-before](https://user-images.githubusercontent.com/2331607/42935727-512a0464-8b42-11e8-9038-724c556bfbab.png)

#### Hover (pr)
![hover-after](https://user-images.githubusercontent.com/2331607/42935789-7333afba-8b42-11e8-9206-0a589a069624.png)
#### Ctrl-hover (pr)
![ctrl-hover-after](https://user-images.githubusercontent.com/2331607/42935799-7c16fe16-8b42-11e8-82c2-b8fcc5f3e127.png)

I find the formatted function signature here much easier to read. But the interaction with definition isn't as good.

## Variable example
#### Hover (before)
![bot-hover-before](https://user-images.githubusercontent.com/2331607/42935876-ae0e5504-8b42-11e8-827e-f31471d3f3e4.png)
#### Ctrl-hover (before)
![bot-ctrl-hover-before](https://user-images.githubusercontent.com/2331607/42935889-b3bd854c-8b42-11e8-8198-ae29f15fbcad.png)


#### Hover (pr)
![bot-hover-after](https://user-images.githubusercontent.com/2331607/42935896-b83e24f0-8b42-11e8-8967-499e0b024c52.png)
#### Ctrl-hover (pr)
![bot-ctrl-hover-after](https://user-images.githubusercontent.com/2331607/42935904-bd4a1904-8b42-11e8-900d-e60a3110a1ef.png)

With variables though I think the previous functionality was better.

So I think maybe our new powers should be used just for improving the types, rather than looking at the definition. But maybe there are other cases?

aloucks commented:
@nrc @alexheretic fallback on racer is now behind the `racer_completion` flag and I've added a separate option to disable showing additional context (e.g. local variable declarations) in the base hover tooltip. Personally I think it's nicer to see those up front rather than having to `ctrl` hover, but the option is now there. I also refactored the tooltip function signatures to make it easier to add additional feature toggles in the future.

aloucks commented:
@nrc @alexheretic 

> Just land this and deal with the fallout when we have to remove Racer (or keep Racer just for docs).

Just to clarify a little - Racer is really only utilized when the `Def` is missing in the save-analysis. This _does_ return the docs (which are then post-processed), but I don't think it's absolutely required.

I'm considering ignoring everything that comes back from racer except for the file path and coordinates and then synthesizing the `Def`. This way, everything else would follow the exact code path as if the `Def` came from the save-analysis.

aloucks commented:
@nrc  @alexheretic 

> I'm considering ignoring everything that comes back from racer except for the file path and coordinates and then synthesizing the Def. This way, everything else would follow the exact code path as if the Def came from the save-analysis.

These changes have been made. Racer really is now _only_ used to resolve a `Def` when it's not available via the save-analysis. Everything else is handled exactly as if it came from the save-analysis. The dependency on racer is really minimal at this point and isn't any greater than the dependency on the goto symbol fallback or completion.  

Modules and crates now have pretty good tooltips (including `std`!). Hovering over the middle components in a module path works now (e.g. hovering over std and sync in `std::sync::Arc` will show docs and module info). Trait bounds on associated types produce tooltips now as well. 

![hover_mod](https://user-images.githubusercontent.com/221559/43041186-7a8f5896-8d26-11e8-8c0c-f495eee5cddc.gif)

Xanewok commented:
@kngwyu you might be interested in https://github.com/rust-lang-nursery/rls/pull/930/commits/85ee3e6208a501611e958facff5db35c5a5c6689#diff-0075e5ccaa51bdbd32798fe7c2fa4914R645

Xanewok commented on src/actions/hover.rs:
```diff
@@ -0,0 +1,2067 @@
+            "
+            Begin first item docs
+
+            The first item docs should not pick
```
It's great to see such an extensive test coverage, that's an awesome thing to have!

While I certainly don't mind having more tests, I'm somewhat concerned this blows up number of currently used test solutions - with this, we'll have integration tests in a form of Cargo unit tests, only a couple of actual Cargo-style integration tests, [additional json tests for diagnostics](https://github.com/rust-lang-nursery/rls/blob/master/src/actions/diagnostics.rs#L341) and now this custom harness for tooltips (which, with test definitions, takes a bit more LOC than the actual logic itself).

It shouldn't block this PR and it's a great thing to have but just as a note, I think we need to take a more streamlined and declarative approach to testing various RLS bits in the future.

Xanewok commented on src/actions/hover.rs:
```diff
@@ -642,9 +642,18 @@ fn racer_match_to_def(ctx: &InitActionContext, m: &racer::Match) -> Option<Def>
     let matchstr_len = matchstr.len() as u32;
     let docs = m.docs.trim().to_string();
     m.coords.map(|coords| {
+        // racer v2.1.0 seems to have a bug where it's populating the row
```
I'd like to see this one solved before merging, especially the 2.1 seems fresh(?) and we can coordinate a fix with @kngwyu rather swiftly

kngwyu commented:
Sorry I fixed racer's coord bag now, @aloucks could you please check it's fixed by change racer dependency to `racer = {version = "2.1.0", git = "https://github.com/racer-rust/racer.git"}` ?

aloucks commented:
@kngwyu the fix on racer/master resolves the issue

aloucks commented on src/actions/hover.rs:
```diff
@@ -0,0 +1,2067 @@
+            file_path,
+            location
+        );
+        let matches = racer::complete_from_file(file_path, location, &session);
```
Good call! At one point it _appeared_ that I was getting better results with `complete_from_file`, but I think there were other things at play. I've updated locally to use `find_definition` and all the tests pass. I'll push the change along with the removal of the zero-index workaround once a new version of racer is available.

kngwyu replied:
Just published racer 2.1.1, including coordinate fix

aloucks commented:
The appveyor breakage is the same failure that's on the the original clippy/racer pull request: https://github.com/rust-lang-nursery/rls/pull/955

It looks like `curl` is failing to build which is causing `libssh2` to fail as well.

```
error: failed to run custom build command for `curl-sys v0.4.6`
...
error: failed to run custom build command for `libssh2-sys v0.2.8`
```

The curl version bump to `0.4.6` had a [failed build](https://ci.appveyor.com/project/alexcrichton/curl-rust/build/1.0.524) but it looks like a different issue (maybe?). There's also some recent curl [commits related to libz and msvc](https://github.com/alexcrichton/curl-rust/commits/master) but it's not clear to me if it's related. 

Oddly enough, I'm able to build locally on Windows.

@alexcrichton would you happen to have any insight to why `curl` is failing to build?

alexcrichton commented:
Ah unfortunately I may not have much insight :(

The error here is:

```
thread 'main' panicked at 'fs::copy(libs.join("lib/libcurl_a.lib"), dst.join("lib/curl.lib")) return the error The system cannot find the path specified. (os error 3)', C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\curl-sys-0.4.6\build.rs:351:5
```

which could happen for any number of reasons, but perhaps path length is one of them? (that's my only real guess though)

alexheretic commented:
Yep we'll need to fix the appveyor build, but it shouldn't block this pr. 

I personally feel this new functionality is quite beneficial, do we all agree on that? Are there outstanding issues blocking the merge?

Xanewok commented:
Yeah, this looks good and is definitely a good addition.

@nrc as I said before, I think this is quite good and since it's quite abstracted away and only falling back on Racer when it can't leverage save analysis, I think we shouldn't worry about this and just merge this now. What do you say?

nrc commented:
Yep, r+

@aloucks could you squash some of the commits please?

aloucks commented:
@nrc I tried squashing it and started running into conflicts. It's probably due to merging in master after rebasing became more tedious as the PR aged. I'm a little nervous at this point now that it's grown quite large and there's been multiple merges in from master.

alexheretic commented:
@aloucks github can squash-merge this, you can also squash it without conflicts with something like:
```
git checkout -b hover2 upstream/master
git merge --squash hover
git commit -m "hover docs ..."
git checkout hover
git reset hover2
```
You can even re-do multiple commits this way if you like.

aloucks commented:
@alexheretic Thanks! 
@nrc should be good now

nrc commented:
Thanks!
